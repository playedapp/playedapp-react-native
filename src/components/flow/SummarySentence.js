import React from "react"
import { Text, StyleSheet } from "react-native"
import PropTypes from "prop-types"
import Colors from "../../constants/Colors"

export const SummarySentence = ({
  games,
  notFollowedPlayers,
  anonymousPlayers,
  onPlayerPress,
  onGamePress,
}) => {
  const players = [
    ...notFollowedPlayers.map(player => (
      <Text
        key={player.key}
        style={styles.link}
        onPress={() => onPlayerPress && onPlayerPress(player)}
      >
        {player.name}
      </Text>
    )),
  ]

  if (anonymousPlayers.length === 1) {
    players.push("one other")
  } else if (anonymousPlayers.length > 1) {
    players.push(`${anonymousPlayers.length} others`)
  }

  return (
    <Text>
      <Text key="game">
        {"Played "}
        <Text
          style={styles.link}
          onPress={() => onGamePress && onGamePress(games[0])}
        >
          {games[0].title}
        </Text>
        {players.length > 0 && <Text> with </Text>}
        {players.reduce((arr, player, index, players) => {
          if (index === 0) {
            return [...arr, player]
          } else if (index === players.length - 1) {
            return [...arr, <Text key={index}> and </Text>, player]
          } else {
            return [...arr, <Text key={index}>, </Text>, player]
          }
        }, [])}
      </Text>
    </Text>
  )
}

const playerList = PropTypes.arrayOf(
  PropTypes.shape({
    key: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  }),
)

SummarySentence.propTypes = {
  games: PropTypes.arrayOf(PropTypes.shape({ title: PropTypes.string }))
    .isRequired,
  notFollowedPlayers: playerList,
  anonymousPlayers: playerList,
  onPlayerPress: PropTypes.func,
  onGamePress: PropTypes.func,
}

SummarySentence.defaultProps = {
  notFollowedPlayers: [],
  anonymousPlayers: [],
}

const styles = StyleSheet.create({
  link: {
    color: Colors.primary,
    fontWeight: "bold",
  },
})
