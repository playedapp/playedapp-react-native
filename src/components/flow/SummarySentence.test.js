/* global test, expect */
import React from "react"
import renderer from "react-test-renderer"
import { SummarySentence } from "./SummarySentence"

const game = {
  id: "123",
  title: "Clans of Caledonia",
  thumbnail: {
    url: "http://192.168.0.4:3000/static/covers/clansofcaledonia.png",
  },
}

const notFollowedPlayer1 = {
  id: "115",
  person: {
    id: "115",
    name: "Nina",
    isFollowedByMe: false,
  },
  score: 27,
  rank: 4,
  ratings: [{ game: "123", currentRating: 4 }],
}

const notFollowedPlayer2 = {
  key: "116",
  person: {
    key: "116",
    name: "Nina",

    isFollowedByMe: false,
  },
  score: 27,
  rank: 4,
  ratings: [{ game: "123", currentRating: 4 }],
}

const notFollowedPlayer3 = {
  key: "114",
  person: {
    key: "114",
    name: "Nina",
    isFollowedByMe: false,
  },
  score: 27,
  rank: 4,
  ratings: [{ game: "123", currentRating: 4 }],
}

const anonymousPlayer = {
  key: 117,
  name: "Player 7",
  score: 2,
  rank: 7,
}

test("Played X", () => {
  const tree = renderer.create(<SummarySentence games={[game]} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with one other", () => {
  const tree = renderer
    .create(
      <SummarySentence games={[game]} anonymousPlayers={[anonymousPlayer]} />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with 2 others", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        anonymousPlayers={[anonymousPlayer, anonymousPlayer]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y and one other", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1]}
        anonymousPlayers={[anonymousPlayer]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y and two others", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1]}
        anonymousPlayers={[anonymousPlayer, anonymousPlayer]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y and Z", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1, notFollowedPlayer2]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y, Z and W", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[
          notFollowedPlayer1,
          notFollowedPlayer2,
          notFollowedPlayer3,
        ]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y, Z and one other", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1, notFollowedPlayer2]}
        anonymousPlayers={[anonymousPlayer]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Played X with Y, Z and two others", () => {
  const tree = renderer
    .create(
      <SummarySentence
        games={[game]}
        notFollowedPlayers={[notFollowedPlayer1, notFollowedPlayer2]}
        anonymousPlayers={[anonymousPlayer, anonymousPlayer]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
