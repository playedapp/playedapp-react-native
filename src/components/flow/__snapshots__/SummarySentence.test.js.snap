// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Played X 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
  </Text>
</Text>
`;

exports[`Played X with 2 others 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    2 others
  </Text>
</Text>
`;

exports[`Played X with Y 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
  </Text>
</Text>
`;

exports[`Played X with Y and Z 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
  </Text>
</Text>
`;

exports[`Played X with Y and one other 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    one other
  </Text>
</Text>
`;

exports[`Played X with Y and two others 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    2 others
  </Text>
</Text>
`;

exports[`Played X with Y, Z and W 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      , 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
  </Text>
</Text>
`;

exports[`Played X with Y, Z and one other 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      , 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    one other
  </Text>
</Text>
`;

exports[`Played X with Y, Z and two others 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      , 
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Nina
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       and 
    </Text>
    2 others
  </Text>
</Text>
`;

exports[`Played X with one other 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Played 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "#4d00ff",
          "fontWeight": "bold",
        }
      }
    >
      Clans of Caledonia
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       with 
    </Text>
    one other
  </Text>
</Text>
`;
